### ISSUE TITLE
Add server_error connection hook

### ISSUE THREAD

**Issue by emanuele-f** [mitmproxy/mitmproxy#5308]
#### Problem Description
Server connection errors like the following are only logged into the console but ignored:

```
[info] 127.0.0.1:38503: error establishing server connection: [Errno 103] Connect call failed (1.2.3.4', 443)
```

After a `server_connect` hook, we have no way to handle these as the connection is terminated and no `server_disconnected`/`tcp_error` hook is called. It would be nice to have an hook to give plugins a possibility to handle these.

#### Proposal
Add a [ConnectionEvents](https://docs.mitmproxy.org/stable/api/events.html#ConnectionEvents) hook which is called when a server error occurs. For consistency with the existent hooks, it could have the following signature:

```
def server_error(self, data: server_hooks.ServerConnectionHookData):
        # A server connection failed (error message in data.server.error)
```

#### Additional context
My workaround with the current API is to manually override `ConnectionHandler.server_event` (https://github.com/emanuele-f/PCAPdroid-mitm/commit/fd891b8e31defb594ef5e585d0b0c9d11465097c)

**Comment by mhils** [comment]
Yes, that sounds good to me! It makes sense to add an event hook for this. 
Would you be willing to give the implementation a stab?

We should also ponder a bit on the event hook name, those are hard to change later. Some more ideas:

- `server_error`: closest to existing terminology, but doesn't make it clear that this only refers to connection establishment errors
- `server_connect_error`: a bit more verbose
- `server_connect_failed`: alternative

I have no super strong opinions, @mitmproxy/devs any thoughts? Unique opportunity for bikeshedding. 😄  🚲🛖

**Comment by emanuele-f** [comment]
> Yes, that sounds good to me! It makes sense to add an event hook for this.

Great!

> Would you be willing to give the implementation a stab?

I suspect this would require so much time wrt the actual value my contribution can bring, to understand the events logic and provide an acceptably clean implementation. I prefer to rely on you :)

**Comment by mhils** [comment]
> I suspect this would require so much time wrt the actual value my contribution can bring, to understand the events logic and provide an acceptably clean implementation. I prefer to rely on you :)

Honestly your monkeypatching looks much more complex than what is needed :)

You can basically copy & adjust ServerConnectedHook ([see code search](https://github.com/mitmproxy/mitmproxy/search?q=ServerConnectedHook&type=code)). And then just 
`await self.handle_hook(server_hooks.ServerErrorHook(hook_data))`
around here:
https://github.com/mitmproxy/mitmproxy/blob/c91685efeb995659a5dd8dfdbbb65be5c55f2ac0/mitmproxy/proxy/server.py#L190

No worries if you don't have the capacity for that, but I won't make any commitments w.r.t. a timeline.

**Comment by mhils** [comment]
All good. I appreciate you opening this issue, just letting you know I won't find the time to tackle it soon. :)

**Comment by mhils** [comment]
All good. I appreciate you opening this issue, just letting you know I
won't tackle it soon. :)

**Comment by haanhvu** [comment]
I'll work on this

**Comment by abbbe** [comment]
I wonder if this approach can eventually allow retrying server connections, potentially to another server address & port?
Would it be possible to return some 'retry' indicator from the error hook, or alter some data structure?

**Comment by errorxyz** [comment]
Fixed in #6806
